File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
     void fun(int ind,vector<int> &ds,vector<vector<int>> &res,vector<int> &can,int target) {
        if(ind == can.size()) {
            if(target == 0) {
                res.push_back(ds);
            }
            return;
        }

        if(can[ind] <= target) {
            ds.push_back(can[ind]);
            fun(ind,ds,res,can,target - can[ind]);
            ds.pop_back();
        }

        fun(ind+1,ds,res,can,target);
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> ds;
        fun(0,ds,res,candidates,target);
        return res;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> answer = new ArrayList<>();
        backtrack(0, target, candidates, new ArrayList<>(), answer);
        return answer;
    }

    private void backtrack(
            int index,
            int target,
            int[] candidates,
            List<Integer> current,
            List<List<Integer>> answer) {

        if (target == 0) {
            answer.add(new ArrayList<>(current));
            return;
        }

        if (index == candidates.length || target < 0) {
            return;
        }

        current.add(candidates[index]);

        backtrack(
            index,
            target - candidates[index],
            candidates,
            current,
            answer
        );

        current.remove(current.size() - 1);
        backtrack(
            index + 1,
            target,
            candidates,
            current,
            answer
        );
    }
}